src/keys.c

```diff
@@ -35,126 +36,112 @@ bool read_bip32_path(buffer_t *buf, bip32_path_t *const out) {
-                                         raw_private_key,
-                                         NULL);
-    }
-
-    if (error == 0) {
-        // new private_key from raw
-        error = cx_ecfp_init_private_key_no_throw(cx_curve, raw_private_key, 32, private_key);
+static inline unsigned int derivation_type_to_derivation_mode(
+    derivation_type_t const derivation_type) {
+    switch (derivation_type) {
+        case DERIVATION_TYPE_ED25519:
+            return HDW_ED25519_SLIP10;
+        case DERIVATION_TYPE_SECP256K1:
+        case DERIVATION_TYPE_SECP256R1:
+        case DERIVATION_TYPE_BIP32_ED25519:
+        default:
+            return HDW_NORMAL;
     }
-
-    explicit_bzero(raw_private_key, sizeof(raw_private_key));
-
-    return error;
 }
 
 /**
- * @brief Derivates public key from private and curve
+ * @brief Converts `signature_type` to `cx_curve`
  *
- * @param derivation_type: curve
- * @param private_key: private key
- * @param public_key public key output
- * @return int: error, 0 if none
+ * @param signature_type: signature_type
+ * @return cx_curve_t: curve result
  */
-static int crypto_init_public_key(derivation_type_t const derivation_type,
-                                  cx_ecfp_private_key_t *private_key,
-                                  cx_ecfp_public_key_t *public_key) {
-    int error = 0;
-    cx_curve_t const cx_curve =
-        signature_type_to_cx_curve(derivation_type_to_signature_type(derivation_type));
+static inline cx_curve_t signature_type_to_cx_curve(signature_type_t const signature_type) {
+    switch (signature_type) {
+        case SIGNATURE_TYPE_SECP256K1:
+            return CX_CURVE_SECP256K1;
+        case SIGNATURE_TYPE_SECP256R1:
+            return CX_CURVE_SECP256R1;
+        case SIGNATURE_TYPE_ED25519:
+            return CX_CURVE_Ed25519;
+        default:
+            return CX_CURVE_NONE;
+    }
+}
 
-    // generate corresponding public key
-    error = cx_ecfp_generate_pair_no_throw(cx_curve, public_key, private_key, 1);
-    if (error != 0) {
-        return error;
+cx_err_t generate_public_key(cx_ecfp_public_key_t *public_key,
+                             bip32_path_with_curve_t const *const path_with_curve) {
+    if ((public_key == NULL) || (path_with_curve == NULL)) {
+        return CX_INVALID_PARAMETER;
     }
 
+    cx_err_t error = CX_OK;
+
+    bip32_path_t const *const bip32_path = &path_with_curve->bip32_path;
+    derivation_type_t derivation_type = path_with_curve->derivation_type;
+    unsigned int derivation_mode = derivation_type_to_derivation_mode(derivation_type);
+    signature_type_t signature_type = derivation_type_to_signature_type(derivation_type);
+    cx_curve_t cx_curve = signature_type_to_cx_curve(signature_type);
+
+    public_key->W_len = 65;
+    public_key->curve = cx_curve;
+
+    // generate corresponding public key
+    CX_CHECK(bip32_derive_with_seed_get_pubkey_256(derivation_mode,
```

Comment:
use bip32_derive_get_pubkey_256 if we dont have seed.
```suggestion
    CX_CHECK(bip32_derive_get_pubkey_256(derivation_mode,
```

Reply:
`bip32_derive_get_pubkey_256` do not have the argument `derivation_mode` that is needed for `ED25519` (`HDW_ED25519_SLIP10`).

I can use `bip32_derive_get_pubkey_256` for the 3 others curves and still use `bip32_derive_with_seed_get_pubkey_256` for `ED25519` but I'm not sure if that is clearer.

Reply:
I think derive_with_seed is fine.